### ISSUE TITLE
API access for data in the LocalChangeEntity table

### ISSUE THREAD

**Issue by ndegwamartin** [google/android-fhir#2048]
**Is your feature request related to a problem? Please describe.**
Currently there's no API to access data in the **LocalChangeEntity** table

**Describe the solution you'd like**
In the client app we'd like a way to obtain _Read Only access_ to the LocalChangeEntity table data, specifically the **SquashedLocalChanges**.

We can add an interface in _FhirEngine.kt_ and the corresponding implementation in _FhirEngineImpl.kt_

```kotlin
//FhirEngine.kt
suspend fun getUnsyncedLocalChanges(): List<SquashedLocalChange>
```

```kotlin
//FhirEngineImpl.kt
override suspend fun getUnsyncedLocalChanges() : List<SquashedLocalChange> = database.getAllLocalChanges()
```

**Describe alternatives you've considered**
N/A

**Additional context**
This will be useful when one wants have the visibility of the _Synced/Unsynced_ resources on the device. 

Here is a sample of what can be achieved by this:

<img src="https://github.com/google/android-fhir/assets/10017086/19fe2e1a-08be-4c0e-a888-8b69da3be327" width="300"/>

**Would you like to work on the issue?**
Yes

**Comment by jingtang10** [comment]
Thanks for raising this. Haven't thought about this carefully, but is it possible at all to reuse the syncUpload API for this purpose? Maybe we could change the API so that if no result is returned from the lambda we do not delete the local changes from the local changes table - so that esesentially you're using that API to read local changes.

**Comment by ndegwamartin** [comment]
It is worth noting that the above functionality/implementation renders the dialog based on a user action rather than at the end of a Sync job

**Comment by ndegwamartin** [comment]
Additional note, we may need to refactor `Database.getAllLocalChanges()` API for optimisation since it reads all the records into memory before performing the squash operation and returning the results. 

For a case where an application using the SDK is offline first, they might have collected a lot of data in the field and its months before they can get a working internet connection to sync upward to the server. 

This might require a new ticket but is worth highlighting here cc @jingtang10

**Comment by jingtang10** [comment]
@anchita-g

**Comment by jingtang10** [comment]
as of the latest design changes in the sync module, we will be deprecating syncUpload and syncDownload methods. the normal sync APIs will still be functioning as usual but we have concluded the syncUpload and syndDownload APIs are not utilised by developers and are overly complex. this actually means we do need to provide sufficient capabilities to manage local changes. so we need to work on this more.

**Comment by santosh-pingle** [comment]
> **Is your feature request related to a problem? Please describe.** Currently there's no API to access data in the **LocalChangeEntity** table
> 
> **Describe the solution you'd like** In the client app we'd like a way to obtain _Read Only access_ to the LocalChangeEntity table data, specifically the **SquashedLocalChanges**.
> 
> We can add an interface in _FhirEngine.kt_ and the corresponding implementation in _FhirEngineImpl.kt_
> 
> ```kotlin
> //FhirEngine.kt
> suspend fun getUnsyncedLocalChanges(): List<SquashedLocalChange>
> ```
> 
> ```kotlin
> //FhirEngineImpl.kt
> override suspend fun getUnsyncedLocalChanges() : List<SquashedLocalChange> = database.getAllLocalChanges()
> ```
> 
> **Describe alternatives you've considered** Might be better to have an API that just returns the counts for the grouped changes per resource because if there's too much data in the LocalChangeEntity table it may lead performance issues since obtaining the counts with the above approach is an in memory operation.
> 
> **Additional context** This will be useful when one wants have the visibility of the _Synced/Unsynced_ resources on the device.
> 
> Here is a sample of what can be achieved by this:
> 
> <img alt="" width="300" src="https://private-user-images.githubusercontent.com/10017086/248945761-19fe2e1a-08be-4c0e-a888-8b69da3be327.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcwNzE4MzgsIm5iZiI6MTcyNzA3MTUzOCwicGF0aCI6Ii8xMDAxNzA4Ni8yNDg5NDU3NjEtMTlmZTJlMWEtMDhiZS00YzBlLWE4ODgtOGI2OWRhM2JlMzI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIzVDA2MDUzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyY2Y4MDg4MzhhNmM2YWUzMjVjOTQ0YmEwMDk3YzI5MTkxYWM1ZDYzNWZjZjk0ZjE0OGIzNzBjMmVjNzlkYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5QkmobpAkfvTCcne5mgygqi5Md8cQWcVVvcuZ_QKGFw">
> **Would you like to work on the issue?** Yes

@ndegwamartin 
//FhirEngine.kt
Could you provide more details on what information is expected in the `SquashedLocalChange` class? 
We could consider including the following:
- Resource type 
- resource IDs
- The total count of local changes per resource
- A list of local change IDs for each resource 
- The count of changes per operation type (e.g., Insert, Update, Delete)

Not sure if the following should be included :
- Local changes per resource type (which could be fetched separately by the application)
- Local changes per operation type and resource type (which can also be fetched separately using change IDs)

**Comment by ndegwamartin** [comment]
@santosh-pingle yeah I think we can enhance the returned results with all the above, but at the very basic level, at least for our use case we just needed the _map_ of Resource type to the grouped(by resourced id) _count_ of all local changes. That is for example, if the same Patient Resource has multiple updates it will show up as one unsynced Patient resource.
